### ISSUE TITLE
Memory leak while submitting tasks. `redis, chain, group`

### ISSUE THREAD

**Issue by PythonFZ** [celery/celery#9475]
# Checklist

- [ ] I have verified that the issue exists against the `main` branch of Celery.
- [ ] This has already been asked to the [discussions forum](https://github.com/celery/celery/discussions) first.
- [ ] I have read the relevant section in the
  [contribution guide](https://docs.celeryq.dev/en/main/contributing.html#other-bugs)
  on reporting bugs.
- [ ] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [ ] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [ ] I have checked the [commit log](https://github.com/celery/celery/commits/main)
  to find out if the bug was already fixed in the main branch.
- [ ] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).
- [ ] I have tried to reproduce the issue with [pytest-celery](https://docs.celeryq.dev/projects/pytest-celery/en/latest/userguide/celery-bug-report.html) and added the reproduction script below.

## Mandatory Debugging Information

- [ ] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [ ] I have verified that the issue exists against the `main` branch of Celery.
The issue can be reproduced with the latest PyPi version as well as `celery-5.5.0rc4 kombu-5.5.0rc2` from `main`.
At least Python `3.11` and `3.12` are affected and I could reproduce it on Ubuntu and MacOS.
- [ ] I have included the contents of ``pip freeze`` in the issue.
- [ ] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- None

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**:

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
❯ celery report

software -> celery:5.5.0rc4 (immunity) kombu:5.5.0rc2 py:3.12.2
            billiard:4.2.1 py-amqp:5.3.1
platform -> system:Darwin arch:64bit
            kernel version:24.1.0 imp:CPython
loader   -> celery.loaders.default.Loader
settings -> transport:amqp results:disabled

deprecated_settings: None
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

- **Minimal Python Version**: N/A or Unknown
- **Minimal Celery Version**: N/A or Unknown
- **Minimal Kombu Version**: N/A or Unknown
- **Minimal Broker Version**: N/A or Unknown
- **Minimal Result Backend Version**: N/A or Unknown
- **Minimal OS and/or Kernel Version**: N/A or Unknown
- **Minimal Broker Client Version**: N/A or Unknown
- **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
❯ pip show redis
Name: redis
Version: 5.2.1
Summary: Python client for Redis database and key-value store
Home-page: https://github.com/redis/redis-py
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
from celery import Celery, chain, group

url = "redis://localhost:6379/5"

app = Celery(
    broker_url=url,
    result_backend=url,
)

@app.task(bind=True, default_retry_delay=5) 
def job1(self,
):
    return 1

levels = 10
nodes_per_level = 25

per_level_groups = []

for _ in range(levels):
    group_tasks = [job1.s(
    ) for _ in range(nodes_per_level)]
    per_level_groups.append(group(group_tasks))

result = chain(per_level_groups).apply_async()
```

</p>
</details>

# Expected Behavior
The chain should be submitted.

# Actual Behavior
The memory usage of the Python process increases until the machine dies or the process gets killed.

**Comment by freddyheppell** [comment]
I am also experiencing an indefinite hang and memory leak when running `apply_async` on a chain of groups with:

- Celery 5.5.0
- RabbitMQ Broker
- django-celery-results 2.6.0 backend

**Comment by auvipy** [comment]
would you mind help us tracing this?

**Comment by freddyheppell** [comment]
I did try attaching a debugger but I ended up just getting lost inside various recursive levels of unfolding/optimising structure and couldn't really tell what was going on. If I had to guess I'd say it's probably an infinite recursion there. 

I'll try it with a profiler on a simpler example and see if I can get anything out of it.

**Comment by auvipy** [comment]
Do you use memray?

**Comment by dfm88** [comment]
Hi guys, I took some time to look into this, I think it is not strictly related to broker/backend configurations, but rather a serialization issue with the JSON serializer in Celery.

**TL;DR**: This does not happen with the `pickle` serializer, only with `JSON`.

The memory leak occurs before being able to submit the task, specifically during the serialization of the chain of groups. Also there is a given threshold of levels (considering the reproducible example) where the memory leak occurs, that made me think this is not a recursion issue, but rather a serialization issue. 

With this configuration memory increase until a certain level (~400 MB in my case), after this threshold it increases exponentially:

```python

@app.task(bind=True, default_retry_delay=5) 
def job1(self, *args, **kwargs):  # takes any args/kwargs to let chain work
    return 1

levels = 3  # with more levels, memory starts to grow exponentially
nodes_per_level = 25
```

I used `memray` and debugger to trace the memory usage and serialization process. Here's what I found:

When the JSON encoder hits nested group objects, it triggers recursive serialization:

```python
# During JSON serialization in kombu/serialization.py:220
ipdb> obj[2]['chord'] 
group([job1('2-1'), job1('2-2'), ..., job1('2-25')])  # 25 nested tasks

# JSON encoder calls into canvas.py
> celery/canvas.py(832)::Signature(dict)::items()
--> 832     def items(self):
    833         for k, v in super().items():
```

In `celery/canvas.py` `_prepare_chain_from_options()`:

```python
elif 'chain' not in options:
    return ChainMap({'chain': tasks}, options)  # ← Passes entire nested structure
```

Here a screenshot of the memray live execution with 10 levels (I forced it to stop after it reached ~2GB of memory):

![Image](https://github.com/user-attachments/assets/995299d0-e4b7-452b-ac96-a71dc959a3a2)

**Comment by auvipy** [comment]
@dfm88 you can check this pr https://github.com/celery/celery/pull/9799 and use the idea from integration test to help reproduce and potentially fix this memory leak issue

**Comment by dfm88** [comment]
> [@dfm88](https://github.com/dfm88) you can check this pr [#9799](https://github.com/celery/celery/pull/9799) and use the idea from integration test to help reproduce and potentially fix this memory leak issue

I created a gist to easily reproduce this (I don't know if it is the case of submitting a pr with this test since it actually doe not solves the problem) 

- GIST https://gist.github.com/dfm88/9646c5a89a1fe52c4c32c3f4924ba416

A way t tun it is

```shell
LEVELS=8 python -m pytest t/integration/test_issue_9475.py::test_json_serialization_memory_leak -v
```

**Comment by auvipy** [comment]
I have opened a PR with your tests
